README.md
# homebridge-inkbird-bt-thsensor
A homebridge-plugin for the Inkbird bluetooth temperature- and humidity-sensors.

### Features:
- Temperatures (with Eve history)
- Humidity (with Eve history)
- Battery level
- Supported sensors:
   - IBS-TH1
   - IBS-TH1-Plus
   - IBS-TH2
   - IBS-TH2-Plus

## Installation:

### 1. Install homebridge and inkbird-bt-thsensor plugin.
- 1.a `sudo npm install -g homebridge --unsafe-perm`
- 1.b `sudo npm install -g homebridge-inkbird-bt-thsensor`
- 1.c ``sudo setcap cap_net_raw+eip $(eval readlink -f `which node`)``

The command 1.c grants the node binary cap_net_raw privileges, so it can start/stop BLE advertising.
Note: The command requires setcap to be installed. It can be installed the following way:
```
    apt: sudo apt-get install libcap2-bin
    yum: su -c \'yum install libcap2-bin\'
```

### 2. Update homebridge configuration file.
```
"accessories": [
   {
      "accessory"             : "InkbirdBtTHSensor",
      "plugin_map"            :
      {
         "plugin_name"        : "homebridge-inkbird-bt-thsensor",
         "index"              : 0
      },
      "name"                  : "Garden TH Sensor",
      "model"                 : "IBS-TH2-Plus",
      "sensor"                : "internal",
      "mac_address"           : "50:51:A9:7D:FC:E9",
      "update_interval"       : 600,
      "offset_int_temperature": -50,
      "offset_ext_temperature": +30,
      "offset_int_humidity"   : +210,
      "storage"               : "filesystem",
      "loglevel"              : 3
   }
]
```

- name                  (required): Choose a suitable name for your sensor accessory.
- model                 (required): Choose a type from list of supported types above.
                                    If your type is not available, but you want to try if your sensor works anyway put
                                    `not in list - try it anyway`
                                    You won't get an error that the sensor is wrong and plausibility and CRC checks will be switched off.
                                    But be warned, you might get very strange values!!!
- sensor                (optional): On supporting sensors (the Plus models), the sensor to be shown can be selected (internal or external). To have both available just put two accessory entries
                                    into your config file
- mac_address           (optional): Put the MAC-address of the sensor if you know it.
                                    If not, leave the value open and the plugin will choose any sensor it finds that passes the plausibility checks. In the log you will get a message like this:
                                    `7/6/2020 12:39:05 [Garden TH Sensor] Peripheral with MAC 50:51:a9:7d:fc:e9 found - stop scanning`
                                    There you have your MAC. Copy it to your configuration in this format ("xx:xx:xx:xx:xx:xx") to lock only to this sensor.
- update_interval       (optional): If you specify an update interval (in seconds) the plugin will automatically refresh the values so you have
                                    a faster response for your value. Also you need to configure this option, if you want the Eve history to be
                                    filled with values. But be advised that this might reduce your batteries lifetime, so don't choose it too short.
- offset_int_temperature(optional): If specified the offset value will be added to the measured internal temperature (value is in units of 0.01°C)
- offset_ext_temperature(optional): If specified the offset value will be added to the measured external temperature (value is in units of 0.01°C)
- offset_int_humidity   (optional): If specified the offset value will be added to the measured internal humidity (value is in units of 0.01%)
- storage               (optional): Where do you want the Eve history to be stored (`filesystem` (default) or `googleDrive` (not tested)).
- loglevel              (optional): The log level at start of the plugin - smaller numbers lead to less messages
                                    (0 = Fatal, 1 = Error, 2 = Warning, 3 = Info (default), 4 = Debug).

### 3. New sensor types
If you have an Inkbird Bluetooth sensor other than the IBS-TH1 please try it out using the `not in list - try it anyway` or the `IBS-TH1` setting for the model parameter. If the sensor works -> fine, if not set your log level to 4 (debug) and you'll see a
debug message somewhat like `Debug - ManufacturerData is 9c077215005be33f08`. Try finding out the protocol by heating up / cooling down the sensor and checking which values change. You can also use the original app to compare the temperature and the ManufacturerData. The protocol of the `IBS-TH1` you find in index.js

And here I'm depending on your help as I only have the `IBS-TH1`. Please let me know everything about new sensors, like:
- Hey, the sensor xyz works with the `IBS-TH1` setting 
- or... the protocol of abc seems to be the following

and I will update the plugin so more people can use it. Thanks a lot 


CHANGELOG.md
# Changelog for homebridge-inkbird-bt-thsensor
This is the change log for the plugin, all relevant changes will be listed here.

For documentation please see the [README](https://github.com/SteidlD/homebridge-inkbird-bt-thsensor/blob/master/README.md)

## 0.5.0
- Support for sensor types IBS-TH2 and IBS-TH2-Plus added
- Selection of internal or external sensor
- Offset for internal, external temperature sensor added
- Offset for internal humidity added

### 0.4.2
- Support for IBS-TH1-Plus

### 0.4.1
- Bugfix Issue #7: Incorrect temperature (problem with negative temperatures)

## 0.4.0
- Actualization also in cyclic mode
- Info internal/external sensor available
- Bugfix Issue #4: Incorrect temperature
- Funktion "not in list - try it anyway" improved
- Debug output for new sensors

### 0.3.1
- Bugfix: Reference error Characteristic

## 0.3.0 (do not use)
- Added Eve history for temperature and relative humidity
- Bug fix cyclic read
- Added log level

## 0.2.0
- Plausibility checks added
- CRC checking added
- Easy-to-configure added (config.schema, start without MAC)

## 0.1.0
- First working version


CRC16_0x18005.js
// Implements the CRC16 algorithm for MODBUS polynomial (0x18005)
//
//-----------------------------------------------------------------------
// Date        Author      Change
//-----------------------------------------------------------------------
// 07.06.2020  D.Steidl    Created
//-----------------------------------------------------------------------

//-----------------------------------------------------------------------
// Global variables
//-----------------------------------------------------------------------

// variables have to be declared explicitly
'use strict'

/** @const {Object} aCRCTable             Table for 8-bit fast calculation */
const aCRCTable   = 
[  0x0000, 0x8005, 0x800F, 0x000A, 0x801B, 0x001E, 0x0014, 0x8011, 0x8033, 0x0036, 0x003C, 0x8039, 0x0028, 0x802D, 0x8027, 0x0022,
   0x8063, 0x0066, 0x006C, 0x8069, 0x0078, 0x807D, 0x8077, 0x0072, 0x0050, 0x8055, 0x805F, 0x005A, 0x804B, 0x004E, 0x0044, 0x8041,
   0x80C3, 0x00C6, 0x00CC, 0x80C9, 0x00D8, 0x80DD, 0x80D7, 0x00D2, 0x00F0, 0x80F5, 0x80FF, 0x00FA, 0x80EB, 0x00EE, 0x00E4, 0x80E1,
   0x00A0, 0x80A5, 0x80AF, 0x00AA, 0x80BB, 0x00BE, 0x00B4, 0x80B1, 0x8093, 0x0096, 0x009C, 0x8099, 0x0088, 0x808D, 0x8087, 0x0082,
   0x8183, 0x0186, 0x018C, 0x8189, 0x0198, 0x819D, 0x8197, 0x0192, 0x01B0, 0x81B5, 0x81BF, 0x01BA, 0x81AB, 0x01AE, 0x01A4, 0x81A1,
   0x01E0, 0x81E5, 0x81EF, 0x01EA, 0x81FB, 0x01FE, 0x01F4, 0x81F1, 0x81D3, 0x01D6, 0x01DC, 0x81D9, 0x01C8, 0x81CD, 0x81C7, 0x01C2,
   0x0140, 0x8145, 0x814F, 0x014A, 0x815B, 0x015E, 0x0154, 0x8151, 0x8173, 0x0176, 0x017C, 0x8179, 0x0168, 0x816D, 0x8167, 0x0162,
   0x8123, 0x0126, 0x012C, 0x8129, 0x0138, 0x813D, 0x8137, 0x0132, 0x0110, 0x8115, 0x811F, 0x011A, 0x810B, 0x010E, 0x0104, 0x8101,
   0x8303, 0x0306, 0x030C, 0x8309, 0x0318, 0x831D, 0x8317, 0x0312, 0x0330, 0x8335, 0x833F, 0x033A, 0x832B, 0x032E, 0x0324, 0x8321,
   0x0360, 0x8365, 0x836F, 0x036A, 0x837B, 0x037E, 0x0374, 0x8371, 0x8353, 0x0356, 0x035C, 0x8359, 0x0348, 0x834D, 0x8347, 0x0342,
   0x03C0, 0x83C5, 0x83CF, 0x03CA, 0x83DB, 0x03DE, 0x03D4, 0x83D1, 0x83F3, 0x03F6, 0x03FC, 0x83F9, 0x03E8, 0x83ED, 0x83E7, 0x03E2,
   0x83A3, 0x03A6, 0x03AC, 0x83A9, 0x03B8, 0x83BD, 0x83B7, 0x03B2, 0x0390, 0x8395, 0x839F, 0x039A, 0x838B, 0x038E, 0x0384, 0x8381,
   0x0280, 0x8285, 0x828F, 0x028A, 0x829B, 0x029E, 0x0294, 0x8291, 0x82B3, 0x02B6, 0x02BC, 0x82B9, 0x02A8, 0x82AD, 0x82A7, 0x02A2,
   0x82E3, 0x02E6, 0x02EC, 0x82E9, 0x02F8, 0x82FD, 0x82F7, 0x02F2, 0x02D0, 0x82D5, 0x82DF, 0x02DA, 0x82CB, 0x02CE, 0x02C4, 0x82C1,
   0x8243, 0x0246, 0x024C, 0x8249, 0x0258, 0x825D, 0x8257, 0x0252, 0x0270, 0x8275, 0x827F, 0x027A, 0x826B, 0x026E, 0x0264, 0x8261,
   0x0220, 0x8225, 0x822F, 0x022A, 0x823B, 0x023E, 0x0234, 0x8231, 0x8213, 0x0216, 0x021C, 0x8219, 0x0208, 0x820D, 0x8207, 0x0202]; 

/** @const {Object} aReverseBitOrder      Table for reversing the bit-order */
const aReverseBitOrder   = [0x0, 0x8, 0x4, 0xC, 0x2, 0xA, 0x6, 0xE, 0x1, 0x9, 0x5, 0xD, 0x3, 0xB, 0x7, 0xF]; 

//-----------------------------------------------------------------------
// Imports
//-----------------------------------------------------------------------

// from JavaScript

// from InkbirdBtThermohumiditySensor

//-----------------------------------------------------------------------
// Classes 
//-----------------------------------------------------------------------

//-----------------------------------------------------------------------
// Exports
//-----------------------------------------------------------------------

/**
 * Function to calculate the CRC-16 with the polynomial MODNUS (0x18005) for a buffer
 * 
 * @param {Buffer} cBuffer                Buffer object with data
 * @param {number} iFrom                  Index of first data byte
 * @param {number} iTo                    Index of last data byte
 * @param {boolean} bInputReflected       Change the order of the 8 input bits
 * @param {boolean} bResultReflected      Change the order of the 16 output bits
 * @param {number} iInitialValue          Initial value to start calculation (usually 0x0000 or 0xFFFF)
 * @param {number} iFinalXORValue         Value that will be XORed with the result (usually 0x0000 or 0xFFFF)
 * @returns {number}                      CRC value (undefined if error)
 */
module.exports = function (cBuffer, iFrom, iTo, bInputReflected, bResultReflected, iInitialValue, iFinalXORValue)
{
   var iCRC = iInitialValue;
   var iIndex;
   var iData;

   // Check parameter on plausibility
   if (((iTo-iFrom+1) > cBuffer.length) || (iTo < iFrom) || (iTo > (cBuffer.length - 1)))
      return undefined;

   for (var iCnt = iFrom; iCnt <= iTo; iCnt++)
   {  // work through all the bytes
      // get byte from buffer
      iData    = cBuffer.readUIntLE(iCnt, 1);
      if (bInputReflected)
      {  // Reverse bitorder on incomming bytes (use table to reverse bitorder in nibbles and swap nibbles)
         iData = (aReverseBitOrder[iData & 0xF] << 4) | (aReverseBitOrder[(iData >> 4) & 0xF])
      }

      // Upper byte CRC xor Data for the index to the table
      iIndex   = iData ^ (iCRC >> 8);
      // New CRC = Table XOR CRC<<8
      iCRC     = (aCRCTable[iIndex] ^ (iCRC << 8)) & 0xFFFF;
   }

   iCRC = ((iCRC ^ iFinalXORValue) & 0xFFFF);
   if (bResultReflected)
   {  // Reverse bitorder on result word (use table to reverse bitorder in nibbles and swap nibbles)
      iCRC = (aReverseBitOrder[ iCRC       & 0xF] << 12) | (aReverseBitOrder[(iCRC >>  4) & 0xF] << 8) | 
             (aReverseBitOrder[(iCRC >> 8) & 0xF] <<  4) | (aReverseBitOrder[(iCRC >> 12) & 0xF]     );
   }
   return iCRC;
}

index.js
// Main file for plugin
//
//-----------------------------------------------------------------------
// Date        Author      Change
//-----------------------------------------------------------------------
// 24.05.2020  D. Steidl   Created
// 04.06.2020  D. Steidl   Implemented
// 07.06.2020  D. Steidl   CRC16 Modbus implemented
// 14.06.2020  D. Steidl   Added Eve history for temperature and relative humidity
// 09.07.2022  D. Steidl   Alternative temperature added
//-----------------------------------------------------------------------

// The Inkbird bluetooth thermo-/hygrometer IBS-TH1 is a bluetooth low energy "peripheral" device. It's using the following protocol:
// - The sensor is sending cyclic advertising frames on channels 37,38,39
// - If the master wants to read out the actual temperature / humidity value it can simply send a scan request
// - The sensor answers with a scan response: 
//    - Device name:
//       o 0x04   (length in bytes)
//       o 0x09   (type of data is device name)
//       o "sps"  (data)
//    - Data:
//       o 0x0A   (length in bytes)
//       o 0xFF   (manufacturer specific)
//       o 0xtttt (temperature in 0,01°C)
//       o 0xhhhh (humidity in 0,01%)
//       o 0xss   (sensor: 0x00 internal sensor, 0x01 external sensor)
//       o 0xcccc (CRC16_MODBUS: polynomial 0x18005, initial value 0xFFFF, input and result reflected, no final xor, data: 0xtt 0xtt 0xhh 0xhh 0xss)
//                (or alternative temperature on the Plus models)
//       o 0xbb   (battery value in %)
//       o 0x08   (no idea what that means - never changes)
// - for configuration or read out of history data a BTLE connection is established
//    - the sensor reports the services / characteristics: 
//       o 0x1800: Generic access (sps)
//       o 0x1801: Generic attibute
//       o 0x180A: Device information
//          o 0x2A23: System ID (0xE9-FC-7D-00-00-A9-51-50)
//          o 0x2A24: Model Number String ("Model Number")
//          o 0x2A25: Serial Number String ("Serial Number")
//          o 0x2A26: Firmware Revision String ("Firmware Revision")
//          o 0x2A27: Hardware Revision String ("Hardware Revision")
//          o 0x2A28: Software Revision String ("1-1")
//          o 0x2A29: Manufacturer Revision String ("INKBIRD")
//          o 0x2A2A: IEEE 11073-20601 Regulatory Certification Data List (0xFE-00-65-78-70-65-72-69-6D-65-6E-74-61-6C)
//          o 0x2A50: PnP ID (0x01-0D-00-00-00-10-01)
//       o 0xFFF0: manufacturer specific
//          o 0xFFF1: cfg data, read-write (0x00-00-00-00 00-00-00-3c 00-00-00-31 2d-37-5a-4b 00-00-00-00)
//          o 0xFFF2: real time data, read-only
//             o 0xtttt (temperature in 0,01°C)
//             o 0xhhhh (humidity in 0,01%)
//             o 0xss   (sensor: 0x00 internal sensor, 0x01 external sensor)
//             o 0xcccc (CRC-16: polynomial 0x18005, initial value 0xFFFF, input and result reflected, no final xor, data: 0xtt 0xtt 0xhh 0xhh 0xss)
//          o 0xFFF3: cfg data 2, read-write
//          o 0xFFF4: measure, read-only
//          o 0xFFF5: recorder frame, read-only
//          o 0xFFF6: history data, notify
//          o 0xFFF7: run/stop recorder, read-write
//          o 0xFFF8: his data type, read-write
//          o 0xFFF9: reset: write-only (06 = reset?)
//
// - entering configuration:
//    - scanning of service/characteristic structure
//    - service 0x1801 turn on indication
//    - read out config and real time data (especially read minimum 9 times 0xFFF2)
//    - read device name
//    - write 06 to 0xFFF9 (reset?)
//
// - reading history
//    - writing cfg data: 0x00-00-00-00 00-00-00-78 00-00-00-31 2d-37-5e-4f 00-00-00-00
//    - read back and check cfg data
//    - write 06 to 0xFFF9 (reset?)

//-----------------------------------------------------------------------
// Global variables
//-----------------------------------------------------------------------

// variables have to be declared explicitly
'use strict'

/** @type {Object}     Pointer to Homebridge.hap.Service */
var cService;
/** @type {Object}     Pointer to Homebridge.hap.Characteristic */
var cCharacteristic;                                                                               
/** @type {Object}     Pointer to Homebridge.hap.uuid */
var cUUIDGen;                                                                                      
/** @type {String}     FW-Version of the plugin (shown in Homekit) */
var strFWVersion;                                                                                  
//-----------------------------------------------------------------------
// Imports
//-----------------------------------------------------------------------

// from JavaScript

// from InkbirdBtThermohumiditySensor
const cInkbirdBtTHSensorAccessory   = require('./InkbirdBtTHSensorAccessory')
const packageJson                   = require('./package.json')
var   cFakeGatoHistoryService;

//-----------------------------------------------------------------------
// Classes 
//-----------------------------------------------------------------------

//-----------------------------------------------------------------------
// Exports
//-----------------------------------------------------------------------

/**
 * Anonymous function called by homebridge
 * 
 * @param {Object} cHomebridge            Pointer to homebridge object
 * @returns {void}                        nothing
 */
module.exports = function (cHomebridge)
{
   // get version from package
   global.strFWVersion              = packageJson.version;
   // require here because of need to call with cHomebridge argument
   global.cFakeGatoHistoryService   = require('fakegato-history')(cHomebridge);

   console.log("Homebridge API version: " + cHomebridge.version + " InkbirdBtTHSensor V" + global.strFWVersion);

   // Service and Characteristic are from hap-nodejs
   global.cService         = cHomebridge.hap.Service;
   global.cCharacteristic  = cHomebridge.hap.Characteristic;
   global.cUUIDGen         = cHomebridge.hap.uuid;
 
   // For platform plugin to be considered as dynamic platform plugin,
   // registerPlatform(pluginName, platformName, constructor, dynamic), dynamic must be true
   cHomebridge.registerAccessory(packageJson.name, "InkbirdBtTHSensor", cInkbirdBtTHSensorAccessory);
}
